lib/fswatcher/fswatcher.go

```diff
@@ -0,0 +1,577 @@
+	w.notifyTimer.Reset(duration)
+}
+
+func (w *watcher) aggregateEvent(evPath string, evTime time.Time, evType eventType) {
+	if evPath == "." || w.rootEventDir.eventCount() == maxFiles {
+		l.Debugln(w, "Scan entire folder")
+		firstModTime := evTime
+		if w.rootEventDir.childCount() != 0 {
+			evType |= w.rootEventDir.eventType()
+			firstModTime = w.rootEventDir.firstModTime()
+		}
+		w.rootEventDir = newEventDir(".", nil)
+		w.rootEventDir.events["."] = &event{
+			path:         ".",
+			firstModTime: firstModTime,
+			lastModTime:  evTime,
+			evType:       evType,
+		}
+		w.resetNotifyTimerIfNeeded()
+		return
+	}
+
+	parentDir := w.rootEventDir
+
+	// Check if any parent directory is already tracked or will exceed
+	// events per directory limit bottom up
+	pathSegments := strings.Split(filepath.ToSlash(evPath), "/")
+
+	// As root dir cannot be further aggregated, allow up to maxFiles
+	// children.
+	localMaxFilesPerDir := maxFiles
+	var currPath string
+	for i, pathSegment := range pathSegments[:len(pathSegments)-1] {
+		currPath = filepath.Join(currPath, pathSegment)
+
+		if event, ok := parentDir.events[currPath]; ok {
+			event.lastModTime = evTime
+			event.evType |= evType
+			l.Debugf("%v Parent %s (type %s) already tracked: %s", w, currPath, event.evType, evPath)
+			return
+		}
+
+		if parentDir.childCount() == localMaxFilesPerDir {
+			l.Debugf("%v Parent dir %s already has %d children, tracking it instead: %s", w, currPath, localMaxFilesPerDir, evPath)
+			w.aggregateEvent(filepath.Dir(currPath),
+				evTime, evType)
+			return
+		}
+
+		// If there are no events below path, but we need to recurse
+		// into that path, create eventDir at path.
+		if _, ok := parentDir.dirs[currPath]; !ok {
+			l.Debugf("%v Creating eventDir: %s", w, currPath)
+			parentDir.dirs[currPath] = newEventDir(currPath, parentDir)
+		}
+		parentDir = parentDir.dirs[currPath]
+
+		// Reset allowed children count to maxFilesPerDir for non-root
+		if i == 0 {
+			localMaxFilesPerDir = maxFilesPerDir
+		}
+	}
+
+	if event, ok := parentDir.events[evPath]; ok {
+		event.lastModTime = evTime
+		event.evType |= evType
+		l.Debugf("%v Already tracked (type %v): %s", w, event.evType, evPath)
+		return
+	}
+
+	childDir, ok := parentDir.dirs[evPath]
+
+	// If a dir existed at path, it would be removed from dirs, thus
+	// childCount would not increase.
+	if !ok && parentDir.childCount() == localMaxFilesPerDir {
+		l.Debugf("%v Parent dir already has %d children, tracking it instead: %s", w, localMaxFilesPerDir, evPath)
+		w.aggregateEvent(filepath.Dir(evPath), evTime, evType)
+		return
+	}
+
+	firstModTime := evTime
+	if ok {
+		firstModTime = childDir.firstModTime()
+		evType |= childDir.eventType()
+		delete(parentDir.dirs, evPath)
+	}
+	l.Debugf("%v Tracking (type %v): %s", w, evType, evPath)
+	parentDir.events[evPath] = &event{
+		path:         evPath,
+		firstModTime: firstModTime,
+		lastModTime:  evTime,
+		evType:       evType,
+	}
```

Comment:
So it seems that we store full path in every path segment, as in the struct looks like:


```
path = <full path>
dirs = {
   <full path>/foo: {
          path: <full path>/foo
          dirs = {
                <full path>/foo/bar: {
                     path: <full path>/foo/bar
                 }
          }
    }
}
```

Same goes for events. I wonder if it wouldn't make more sense just to store the child name, rather than full path to the child.